### PR TITLE
Moved THTensor_(cauchy) to ATen

### PR DESCRIPTION
**Pull request by moon05**
I think the headings says it all. Will update if more description is necessary.

### CODE REVIEW

File: aten/src/ATen/Declarations.cwrap [hunk 1]
```diff
@@ -2676,7 +2676,6 @@
   types:
```
Comment (syed-ahmed) [inline aten/src/ATen/Declarations.cwrap]: Once https://github.com/pytorch/pytorch/pull/20622 is in, remove this whole block: 
```
[[
  name: _th_cauchy_
  types:
    - floating_point
  backends:
    - CPU
  cname: cauchy
  variants: function
  return: self
  arguments:
    - THTensor* self
    - arg: THGenerator* generator
      default: nullptr
      kwarg_only: True
    - arg: double median
      default: 0
    - arg: double sigma
      default: 1
]]
```

File: aten/src/ATen/native/LegacyDefinitions.cpp [hunk 2]
```diff
@@ -54,6 +54,11 @@ Tensor argsort(const Tensor & self, int64_t dim, bool descending) {
   return std::get<1>(at::sort(self, dim, descending));
 }
 
+Tensor & cauchy_cuda_(Tensor& self, double median, double sigma, Generator * generator) {
+  return at::legacy::th::_th_cauchy_(self, median, sigma, generator);
```
Comment (syed-ahmed) [inline aten/src/ATen/native/LegacyDefinitions.cpp]: Remove this function when this is in: https://github.com/pytorch/pytorch/pull/20622

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -3181,11 +3181,11 @@
     CPU: legacy::cpu::_th_normal_
     CUDA: legacy::cuda::_th_normal_
 
-- func: cauchy_(Tensor(a!) self, float median=0, float sigma=1, *, Generator? generator=None) -> Tensor(a!)
+  - func: cauchy_(Tensor(a!) self, float median=0, float sigma=1, *, Generator? generator=None) -> Tensor(a!)
```
Comment (syed-ahmed) [inline aten/src/ATen/native/native_functions.yaml]: there is an extra space in front of the `- func`

**Review by syed-ahmed — CHANGES_REQUESTED** [review]
Wait for this to go in: https://github.com/pytorch/pytorch/pull/20622. Otherwise LGTM :)

**Comment by ezyang** [comment]
Obsoleted by @syed-ahmed's port,

**Comment by syed-ahmed** [comment]
Re-opening because this is porting the CPU side from TH to ATen. I didn't do that in my PR.

**Review by syed-ahmed — APPROVED** [review]

**Review by zou3519 — CHANGES_REQUESTED** [review]
Could we see some before/after performance numbers to ensure there isn't a performance regression?

File: aten/src/ATen/native/Distributions.cpp [hunk 4]
```diff
@@ -171,6 +171,17 @@ Tensor& bernoulli_scalar_cpu_(Tensor& self, double p, Generator* gen) {
   return self;
 }
 
+  Tensor& cauchy_cpu_(Tensor& self, double median, double sigma, Generator* gen) {
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "cauchy_cpu", [&] {
+    THGenerator* generator = get_generator(gen);
+    std::lock_guard<std::mutex> lock(generator->mutex);
+    CPU_tensor_apply1<scalar_t>(
```
[the remark this answers is not in the document]
Comment (fmassa) [reply aten/src/ATen/native/Distributions.cpp]: Shouldn't we be preferably using TensorIterator here instead?

[the remark this answers is not in the document]
Comment (VitalyFedyunin) [reply aten/src/ATen/native/Distributions.cpp]: To be honest we havn't yet tested parallel calls to random generators. `CPU_tensor_apply1` here does serial processing.

**Comment by VitalyFedyunin** [comment]
Implements #24684

**Comment by ezyang** [comment]
@moon05 Do you still have time to work on this PR?

**Pull request closed without merging**